### ISSUE TITLE
FRAME: Intent Functions 

### ISSUE THREAD

**Issue by kianenigma** [paritytech/polkadot-sdk#5958]
Related to https://github.com/paritytech/polkadot-sdk/issues/216 / https://forum.polkadot.network/t/wasm-view-functions/1045

## TLDR

- A `view_function` is an abstraction over `#[pallet::storage]`.
- An `intent_function` is an abstraction over `#[pallet::call]`.
- Both are meant to be stable, and fill the gap between what the (low level) runtime coder thinks, and what the (high level) DApp developer needs to deliver to end users.  

## Intent

Let's define a "user intent" as a high level, intuitive operation that the user (DOT holder) wants to do. Examples: 

1. Buy coretime with the market price in the on-demand market, and immediately use it to produce one block.
2. Unlock all available governance locks
3. Submit this governance/spend proposal (all the steps that `opengov-cli` tells you)
4. Stake 200 DOTs (and you figure out if I should join a pool or be a direct staker)

Intents are a common word in blockchains, and baking them into FRAME can make Substrate based (para)chain apt in being "intent-centric" ([1](https://sedaprotocol.medium.com/what-are-blockchain-intents-with-seda-a64e2274e1e7), [2](https://www.coindesk.com/tech/2023/11/15/intents-are-blockchains-big-new-buzzword-what-are-they-and-what-are-the-risks/)). 

## Context 

Pallets expose dispatchble functions within `#[pallet::call]`. Such calls are meant to be an atomic transaction, do one thing and do them right, and be easy to weigh, such that their worst case estimate is not that far from the final weight consumed. 

This design for dispatchables is ideal for the low level runtime: it cares about efficiency and correctness. But, it is and inconvenient for application developers.

The runtime's only assistance for this issue is `utility.batch`, assuming the DApp developers will know which calls to batch together to execute an intent. 

In other words, the issue is as follows: 

- The runtime provides (ideally) very low level dispatchable functions. Runtime developers are not expected to think about, or optimize for high level user intent. 
- DApp developers deal with user intentions. And these intentions are _miles apart_ from the underlying dispatchables, and therefore difficult to connect to one another 

```
User intention <--------------------------- 👨‍💻😭 ---------------------------> Runtime Dispatchable
```

## Problem 

The concerete problems are then as follows: 

1. Many developers might have a hard time fuguring out how to map an "intent" to the right sequence of calls that need to be dispatched. 
2. Anecdotally we have learned that very often developers are not encouraged to pursue implementing intents at all, and instead reside to exposing the low level calls to users. This is why we sometimes see DApps that has one button per `#[pallet::call]`.

## Solution

Instead of trying to make `#[pallet::call]` cater for the two groups, introduce a new concept, intent functions specifically to centered around user intentions, rather than being optimized for STF. 

Retain the assumption that a good dispachable should adhere to the above principles only (simple, atomic, weigh-able). Encourage all pallets to expose more high level dispatchables as intent functions. This helps with writing correct and accurately benchmarked code. 

The most simple implementation that comes to mind is to make intent functions a special `view_function` that returns a `RuntimeCall` as a result. This `RuntimeCall` (quite possibly being a `utility.batch`) will fulfills an intent, and can be signed and sent by the user. 

This implies that intent functions are also meant to be stable. 

This implies that intent functions can also be seen (at least) in two levels: pallet-level and runtime-level

### Weight 

One benefit of keeping transactions small and atomic is being accurately weigh-able. You might say this is not a big deal, as the weight is a "worst case estimate", and the `proof_size` is always refunded no matter the benchmarking. But, note that a transaction cannot be dispatched if the worst case weight fee cannot be guaranteed. This can make certain operations inaccessible to small DOT holders. 

> I have heard the anecdote that many small stakers cannot delegate, because the delegation fee estimate is too high. 

### Examples

Some pallets indeed provide dispatchables that are only there as a stept towards intents: 
- `Proxy.remove_proxies`: could have been a batch of `remove_proxy`
- I believe a number of extrinsics in `pallet-xcm` are designed around common user intentions. 

I cannot say I know the code of all pallets by heart, but my guess is that some pallets take more steps towards intents than others', based on the author's preference. A standardized solution will encourage a more unified thinking across all pallets. 

### Feature Gating

Both view functions and intent functions have one common property in common: _**they are not needed at all for the blockchain state transition function**_.

This means that one common pushback towards adding both of them is "it is not needed as a part of STF, it should be implemented offchain, it will increase the WASM size".

Yet, I strongly think that the [3 reasons enumerated here](https://github.com/polkadot-api/polkadot-api/issues/666#issuecomment-2321095016) are good enough reasons to start introducing these "non-STF critical" code into the WASM blob. In the long run, we can explore some kind of "feature gating", where we either maintain two copies of a wasm (stf-only WASM in `:code` that contains a the hash of a secondary `:code_aux`), or the XCQ method where the code for execution of the query has to be sent with the query itself.

**Comment by bkchr** [comment]
Yes, intent makes sense, but they are also just a new kind of buzzword. Generally, I'm still not convinced that this should be part of the wasm. This makes just zero sense, because there can be millions of intents and I have no idea how you want to put these into one runtime. Also what you are forgetting is that you want to execute stuff across different chains. Just being able to "solve" your own chain local intents is not really a good solution.

IMO what you want is a chatgpt trained on all the parachains in Polkadot that can actually solve intents like "Buy coretime with the market price in the on-demand market, and immediately use it to produce one block".

**Comment by kianenigma** [comment]
> Generally, I'm still not convinced that this should be part of the wasm. This makes just zero sense, because there can be millions of intents and I have no idea how you want to put these into one runtime

WDYT about the feature gating part? 

Both view functions and intent functions suffer from this; they are part of the WASM (among other things, like genesis and OCW) that don't contribute to STF.

> Also what you are forgetting is that you want to execute stuff across different chains. Just being able to "solve" your own chain local intents is not really a good solution.

I think intent functions make sense if you agree view functions make sense to exist; they are "one step" toward a better abstraction, not all the way. We need further steps on top of it indeed.

**Comment by bkchr** [comment]
> Both view functions and intent functions suffer from this; they are part of the WASM (among other things, like genesis and OCW) that don't contribute to STF.

My point was not that they don't contribute to STF, but that there are millions of intents and structuring this off-chain is much more reasonable IMO.



> I think intent functions make sense if you agree view functions make sense to exist; they are "one step" toward a better abstraction, not all the way. We need further steps on top of it indeed.

View functions are a little bit more reasonable to some extend. Like abstracting "what are my locks" or whatever, but you can also not write view functions for everything. IMO it still makes more sense to have these abstractions in a library that calls into the runtime.
